### PR TITLE
Add circom highlights

### PR DESCRIPTION
**Pull request by alexandr-martirosyan**

### CODE REVIEW

File: lua/nvim-treesitter.lua [hunk 1]
```diff
@@ -4,6 +4,9 @@ local info = require "nvim-treesitter.info"
 local configs = require "nvim-treesitter.configs"
 local statusline = require "nvim-treesitter.statusline"
 
+-- Ensure Neovim treats .circom files as the "circom" filetype
+vim.cmd "autocmd BufRead,BufNewFile *.circom set filetype=circom"
```
Comment (lucario387) [inline lua/nvim-treesitter.lua:8]: Sorry but this requires a PR to vim first to add the filetype. Please don't add it here

File: lua/nvim-treesitter/parsers.lua [hunk 2]
```diff
@@ -2650,6 +2650,15 @@ list.templ = {
   maintainers = { "@vrischmann" },
 }
 
+list.circom = {
+  install_info = {
+    url = "https://github.com/Decurity/tree-sitter-circom.git",
+    files = { "src/parser.c" },
+    branch = "main",
```
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:2657]:
```suggestion
```

Only needed if it's not the default branch.

And please make sure that 

1. `tree-sitter build`
2. `tree-sitter generate`

works without errors and warnings.

**Review by clason — CHANGES_REQUESTED** [review]
Can you also add injections (at least for `comment`) and folds?

File: lua/nvim-treesitter.lua (hunk 1 above)
Comment (alexandr-martirosyan) [reply lua/nvim-treesitter.lua:8]: ok

File: lua/nvim-treesitter/parsers.lua (hunk 2 above)
Comment (alexandr-martirosyan) [reply lua/nvim-treesitter/parsers.lua]: ok

File: queries/circom/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+  "==>"
+  "<--"
+  "-->"
+  "==="
+] @operator
+
+; Comments
+(comment) @comment @spell
+
+((comment) @comment.documentation
```
Comment (clason) [inline queries/circom/highlights.scm:126]: I would recommend against this; instead change the grammar to capture docstring comments as a separate node.

(And I meant "add an `injections.scm` query", not additional highlights, for comments.)

**Comment by alexandr-martirosyan** [comment]
> Can you also add injections (at least for `comment`) and folds?

I am not a creator of [tree-sitter-circom](https://github.com/Decurity/tree-sitter-circom), I just modified some `.scm` files to be compatible with `nvim-treesitter` plugin

**Comment by clason** [comment]
You can still create an `injections.scm`; look at other languages here.

File: lua/nvim-treesitter/parsers.lua (hunk 2 above)
Comment (alexandr-martirosyan) [reply lua/nvim-treesitter/parsers.lua]: after removing branch = "main" below error occurred

```
mv: cannot stat 'tree-sitter-circom-tmp/tree-sitter-circom-master': No such file or directory
Error detected while processing command line:
Failed to execute the following command:
{
  cmd = "mv",
  opts = {
    args = { "-f", "tree-sitter-circom-tmp/tree-sitter-circom-master", "tree-sitter-circom" },
    cwd = "/home/runner/.local/share/nvim"
  }
}
```

File: lua/nvim-treesitter/parsers.lua [hunk 4]
```diff
@@ -2650,6 +2650,14 @@ list.templ = {
   maintainers = { "@vrischmann" },
 }
 
+list.circom = {
```
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:2653]: Please keep the list alphabetized, and **don't** add the `.git`.

File: lua/nvim-treesitter/parsers.lua (hunk 2 above)
Comment (alexandr-martirosyan) [reply lua/nvim-treesitter/parsers.lua]: @clason

Comment (clason) [reply lua/nvim-treesitter/parsers.lua]: You need to add a lockfile.json entry for the new parser, too (again, keeping the list alphabetical).

File: queries/circom/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+; Definitions
+; -----------
+(function_definition
+  name: (identifier) @function)
+
+(template_definition
+  name: (identifier) @function)
+
+; Use contructor coloring for special functions
+(main_component_definition) @constructor
```
Comment (ribru17) [inline queries/circom/highlights.scm:28]: This whole thing shouldn't be constructor, perhaps just `"main"`?

**Comment by alexandr-martirosyan** [comment]
@ribru17 done!

**Comment by alexandr-martirosyan** [comment]
@ribru17 done!

File: queries/circom/folds.scm [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+  (template_body)
+  (block_statement)
+  (if_statement)
+  (for_statement)
+  (while_statement)
+  (function_body)
+  (call_expression)
+  (array_expression)
+  (tuple_expression)
+  (comment)
```
[the remark this answers is not in the document]
Comment (ribru17) [reply queries/circom/folds.scm:12]: @alexandr-martirosyan sorry, actually the plus sign is necessary here (so it folds multiple import statements as one fold), also would you mind putting it as the last item? There is a weird bug around the ordering with quantified patterns that I am looking into...

[the remark this answers is not in the document]
Comment (alexandr-martirosyan) [reply queries/circom/folds.scm:11]: @ribru17  Thanks for info. Done!

**Review by ribru17 — APPROVED** [review]
Thanks!

**Comment by alexandr-martirosyan** [comment]
I accidentally closed PR, sorry :( Opening again

**Pull request merged**